# ❤️ Phase 7: Heart Disease Risk Assessment

## Objective

The objective of this phase is to enhance the heart disease prediction system by providing a heart disease risk score and corresponding risk level in addition to the prediction result.

Instead of only predicting whether a patient has heart disease, the optimized XGBoost model will estimate the probability of heart disease using `predict_proba()`. This probability will then be converted into a percentage-based risk score and classified into different risk categories to improve interpretability for healthcare professionals and patients.

---

## Tasks Performed

- Load the optimized XGBoost model.
- Prepare patient data for prediction.
- Predict whether the patient has heart disease.
- Generate the probability of heart disease using `predict_proba()`.
- Convert the probability into a percentage-based risk score.
- Classify the patient into an appropriate risk category.
- Display the prediction, risk score, and risk level.

---

## Risk Categories

| Risk Score | Risk Level |
|------------|------------|
| **0–39%** | 🟢 Low Risk |
| **40–69%** | 🟡 Moderate Risk |
| **70–100%** | 🔴 High Risk |

---

## Expected Outputs

- Heart disease prediction
- Heart disease risk score (0–100%)
- Risk category (Low, Moderate, or High)
- Patient-friendly prediction report

## Observation

In this phase, the optimized XGBoost model will be used to predict whether a patient has heart disease and estimate the probability of the disease. The predicted probability will be converted into a risk percentage and categorized into Low, Moderate, or High Risk, making the prediction easier to interpret and more useful for clinical decision-making.

In [1]:
# ==========================================================
# Step 1: Import Required Libraries
# ==========================================================

import pandas as pd
import numpy as np
import joblib

print("Libraries imported successfully.")

Libraries imported successfully.


In [2]:
# ==========================================================
# Step 2: Load the Optimized XGBoost Model
# ==========================================================

optimized_model = joblib.load("../models/optimized_xgboost.pkl")

print("Optimized XGBoost model loaded successfully.")

Optimized XGBoost model loaded successfully.


## Observation

The optimized XGBoost model was successfully loaded from the **models** directory. This trained model will be used throughout this phase to predict the presence of heart disease, estimate the probability of disease, calculate the patient's risk score, and determine the corresponding risk category.

In [3]:
# ==========================================================
# Step 3: Create Patient Data for Prediction
# ==========================================================

patient_data = pd.DataFrame({

    "gender": [2],          # 1 = Female, 2 = Male
    "height": [170],        # cm
    "weight": [78],         # kg
    "ap_hi": [145],         # Systolic BP
    "ap_lo": [90],          # Diastolic BP
    "cholesterol": [2],     # 1=Normal, 2=Above Normal, 3=Well Above Normal
    "gluc": [1],            # 1=Normal, 2=Above Normal, 3=Well Above Normal
    "smoke": [0],           # 0=No, 1=Yes
    "alco": [0],            # 0=No, 1=Yes
    "active": [1],          # 0=No, 1=Yes
    "age_years": [55],
    "bmi": [26.99]

})

print("Patient data created successfully.\n")

patient_data

Patient data created successfully.



,gender,height,weight,ap_hi,ap_lo,cholesterol,gluc,smoke,alco,active,age_years,bmi
0,2,170,78,145,90,2,1,0,0,1,55,26.99


## Observation

A sample patient record was created using the same set of features that were used to train the optimized XGBoost model. The patient data includes demographic information, clinical measurements, lifestyle factors, and Body Mass Index (BMI).

This patient record will serve as the input for generating the heart disease prediction and calculating the corresponding risk score.

In [5]:
# ==========================================================
# Step 4: Generate Heart Disease Prediction
# ==========================================================

# Predict heart disease
prediction = optimized_model.predict(patient_data)[0]

# Convert prediction into readable format
if prediction == 1:
    prediction_result = "Heart Disease Detected"
else:
    prediction_result = "No Heart Disease"

print("Prediction completed successfully.\n")

print(f"Prediction Result: {prediction_result}")

Prediction completed successfully.

Prediction Result: Heart Disease Detected


## Observation

The optimized XGBoost model successfully predicted the heart disease status for the given patient record.

Based on the provided clinical and lifestyle features, the model classified the patient as **Heart Disease Detected**. This prediction will now be supplemented with a probability-based risk score and corresponding risk category to provide a more informative risk assessment.

In [6]:
# ==========================================================
# Step 5: Calculate Heart Disease Risk Score
# ==========================================================

# Predict probabilities
probabilities = optimized_model.predict_proba(patient_data)

# Probability of Heart Disease (Class = 1)
risk_probability = probabilities[0][1]

# Convert to percentage
risk_score = risk_probability * 100

print("Risk score calculated successfully.\n")

print(f"Probability of No Heart Disease : {probabilities[0][0]:.4f}")
print(f"Probability of Heart Disease    : {probabilities[0][1]:.4f}")

print(f"\nHeart Disease Risk Score : {risk_score:.2f}%")

Risk score calculated successfully.

Probability of No Heart Disease : 0.2646
Probability of Heart Disease    : 0.7354

Heart Disease Risk Score : 73.54%


## Observation

The optimized XGBoost model successfully calculated the probability of heart disease for the given patient.

The predicted probability of heart disease was **73.54%**, indicating a relatively high likelihood of the disease. This probability was converted into a percentage-based risk score, which will be used in the next step to classify the patient into the appropriate risk category.

In [7]:
# ==========================================================
# Step 6: Classify Patient Risk Level
# ==========================================================

if risk_score < 40:
    risk_level = "🟢 Low Risk"

elif risk_score < 70:
    risk_level = "🟡 Moderate Risk"

else:
    risk_level = "🔴 High Risk"

print("Risk category assigned successfully.\n")

print(f"Risk Score : {risk_score:.2f}%")
print(f"Risk Level : {risk_level}")

Risk category assigned successfully.

Risk Score : 73.54%
Risk Level : 🔴 High Risk


## Observation

The patient's risk score was successfully classified into a predefined risk category.

Based on the calculated heart disease risk score of **73.54%**, the patient was classified as **🔴 High Risk**. This indicates a high probability of heart disease and suggests that the patient may require further medical evaluation and timely clinical intervention.

In [8]:
# ==========================================================
# Step 7: Display Final Heart Disease Risk Assessment Report
# ==========================================================

print("=" * 60)
print("        HEART DISEASE RISK ASSESSMENT REPORT")
print("=" * 60)

print(f"Prediction        : {prediction_result}")
print(f"Risk Score        : {risk_score:.2f}%")
print(f"Risk Level        : {risk_level}")

print("=" * 60)
print("Assessment completed successfully.")
print("=" * 60)

        HEART DISEASE RISK ASSESSMENT REPORT
Prediction        : Heart Disease Detected
Risk Score        : 73.54%
Risk Level        : 🔴 High Risk
Assessment completed successfully.


## Observation

A comprehensive heart disease risk assessment report was generated using the optimized XGBoost model.

The report presents the prediction result, the calculated heart disease risk score, and the corresponding risk category in a clear and interpretable format. This enhances the usability of the model by providing both a classification outcome and a quantitative assessment of cardiovascular risk.

In [9]:
# ==========================================================
# Step 8: Create Heart Disease Risk Assessment Function
# ==========================================================

def assess_patient(patient_data):
    """
    Predict heart disease, calculate risk score,
    and determine risk category.
    """

    # Prediction
    prediction = optimized_model.predict(patient_data)[0]

    # Prediction label
    if prediction == 1:
        prediction_result = "Heart Disease Detected"
    else:
        prediction_result = "No Heart Disease"

    # Probability
    probability = optimized_model.predict_proba(patient_data)[0][1]

    # Risk Score
    risk_score = probability * 100

    # Risk Category
    if risk_score < 40:
        risk_level = "🟢 Low Risk"
    elif risk_score < 70:
        risk_level = "🟡 Moderate Risk"
    else:
        risk_level = "🔴 High Risk"

    # Final Report
    print("=" * 60)
    print("      HEART DISEASE RISK ASSESSMENT REPORT")
    print("=" * 60)
    print(f"Prediction : {prediction_result}")
    print(f"Risk Score : {risk_score:.2f}%")
    print(f"Risk Level : {risk_level}")
    print("=" * 60)

    return {
        "Prediction": prediction_result,
        "Risk Score": round(risk_score, 2),
        "Risk Level": risk_level
    }

## Observation

A reusable function named `assess_patient()` was developed to automate the heart disease risk assessment process.

The function accepts a patient's clinical data as input and performs heart disease prediction, calculates the probability-based risk score, determines the corresponding risk category, and generates a comprehensive assessment report.

This modular implementation improves code reusability, readability, and makes the project more suitable for future deployment as a web application or API.

# 📌 Phase 7 Summary

## Objective Achieved

The objective of this phase was to enhance the heart disease prediction system by providing a quantitative risk score and corresponding risk category in addition to the prediction result.

## Tasks Completed

- Imported the required libraries.
- Loaded the optimized XGBoost model.
- Created a sample patient record for prediction.
- Predicted the presence of heart disease.
- Calculated the probability of heart disease using `predict_proba()`.
- Converted the probability into a percentage-based risk score.
- Classified the patient into the appropriate risk category.
- Generated a comprehensive heart disease risk assessment report.
- Developed a reusable `assess_patient()` function to automate the complete risk assessment process.

## Final Results

| Output | Result |
|---------|--------|
| Prediction | **Heart Disease Detected** |
| Risk Score | **73.54%** |
| Risk Level | **🔴 High Risk** |

## Conclusion

The optimized XGBoost model successfully predicted the patient's heart disease status and generated a probability-based risk assessment. By converting the prediction probability into a percentage-based risk score and categorizing it into predefined risk levels, the system provides a more interpretable and clinically meaningful assessment.

Additionally, a reusable `assess_patient()` function was developed to automate the entire prediction workflow. This modular implementation improves code reusability, maintainability, and makes the project ready for future deployment as a web application or API.